### PR TITLE
[DEV-2933] payment quotes

### PR DESCRIPTION
**Pull request by bernd2022**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/subdomains/core/payment-link/dto/payment-link.dto.ts [hunk 1]
```diff
@@ -1,39 +1,39 @@
...
+
 export interface PaymentLinkPayRequestDto {
   tag: string;
   callback: string;
   minSendable: number;
   maxSendable: number;
   metadata: string;
-  transferAmounts: TransferInfo[];
+  quoteId: string;
+  requestedAmount: RequestedAmountAsset;
```
Comment (davidleomay) [inline src/subdomains/core/payment-link/dto/payment-link.dto.ts:35]: Maybe like this?
```ts
quote: {
  id: string,
  expiration: Date,
}
```

File: src/subdomains/core/payment-link/services/payment-quote.service.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    assets: Asset[],
+    payment: PaymentLinkPayment,
+  ): Promise<TransferAmount> {
+    const gasPrice = this.evmGasPriceService.getGasPrice(blockchain);
+
+    const transferAmount: TransferAmount = {
+      method: blockchain,
+      gasPrice: gasPrice,
+      assets: [],
+    };
```
[the remark this answers is not in the document]
Comment (davidleomay) [reply src/subdomains/core/payment-link/services/payment-quote.service.ts:166]: gas price in wei?

File: src/subdomains/core/payment-link/dto/payment-link.dto.ts (hunk 1 above)
Comment (davidleomay) [reply src/subdomains/core/payment-link/dto/payment-link.dto.ts:35]: Add `displayName`

File: src/subdomains/generic/forwarding/services/lnurl-forward.service.ts [hunk 3]
```diff
@@ -55,7 +59,12 @@ export class LnUrlForwardService {
...
       maxSendable: mSatTransferAmount.amount,
       metadata: LightningHelper.createLnurlMetadata(pendingPayment.requestMemo),
-      transferAmounts: pendingPayment.transferInfo,
+      quoteId: actualQuote.uniqueId,
+      requestedAmount: {
+        asset: pendingPayment.currency.name,
+        amount: pendingPayment.amount,
+      },
+      transferAmounts: actualQuote.transferAmountsAsObj,
     };
```
Comment (davidleomay) [inline src/subdomains/generic/forwarding/services/lnurl-forward.service.ts:68]: Filter out `MSAT` amount

File: src/integration/blockchain/shared/evm/evm-gas-price.service.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    await this.updateGasPrice();
+  }
+
+  private async updateGasPrice() {
+    for (const blockchain of EvmGasPriceService.GAS_PRICE_BLOCKCHAINS) {
+      try {
+        this.gasPriceCache.set(blockchain, {
+          timestamp: new Date(),
+          gasPrice: EvmUtil.fromWeiAmount(await this.evmRegistryService.getClient(blockchain).getRecommendedGasPrice()),
+        });
```
[the remark this answers is not in the document]
Comment (bernd2022) [reply src/integration/blockchain/shared/evm/evm-gas-price.service.ts:56]: Shouldn't it be “gwei”?

File: src/subdomains/core/payment-link/dto/payment-link.dto.ts (hunk 1 above)
Comment (bernd2022) [reply src/subdomains/core/payment-link/dto/payment-link.dto.ts:35]: What do you mean with "Add displayName"?

File: src/subdomains/generic/forwarding/services/lnurl-forward.service.ts [hunk 5]
```diff
@@ -95,16 +101,16 @@ export class LnUrlForwardService {
...
+    const asset = isBtc ? 'BTC' : params.asset;
     const method = Util.toEnum(Blockchain, params.method) ?? Blockchain.LIGHTNING;
 
     return {
       method: method,
       asset: asset,
-      amount: isMsat ? LightningHelper.msatToBtc(amount) : amount,
+      amount: isBtc ? LightningHelper.msatToBtc(amount) : amount,
       quoteUniqueId: params.quote,
     };
```
Comment (davidleomay) [inline src/subdomains/generic/forwarding/services/lnurl-forward.service.ts:115]: Does this make sense? If the asset is BTC, convert from MSAT to BTC? I think we can leave this as is, or just remove `|| params.asset === 'MSAT'` (line 98/104)

File: src/integration/blockchain/shared/evm/evm-gas-price.service.ts [hunk 6]
```diff
@@ -61,6 +60,10 @@ export class EvmGasPriceService implements OnModuleInit {
     }
   }
 
+  private convertGasPrice(amount: BigNumber): number {
+    return parseFloat(ethers.utils.formatUnits(amount, 'wei'));
+  }
```
Comment (davidleomay) [inline src/integration/blockchain/shared/evm/evm-gas-price.service.ts:65]: Isn't it just `+amount`?

**Comment by TaprootFreak** [comment]
👀

**Review by davidleomay — APPROVED** [review]
Wait for [[DEV-2933]](https://github.com/DFXswiss/api/pull/1533)

**Pull request merged**